File: android-app/app/src/main/java/com/agnes/editimage/ui/EditImageViewModel.kt
package com.agnes.editimage.ui

import android.app.Application
import android.net.Uri
import android.os.SystemClock
import android.util.Base64
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.agnes.editimage.BuildConfig
import com.agnes.editimage.data.AgnesApi
import com.agnes.editimage.data.SKILLS
import com.agnes.editimage.data.analyzeAndEnhance
import com.agnes.editimage.data.generateEdit
import com.agnes.editimage.data.pickRatio
import com.agnes.editimage.util.imageDimensions
import com.agnes.editimage.util.readBytes
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class LoadedSkill(val badge: String, val chars: Int)

sealed interface ChatItem {
    data class UserMessage(val text: String, val imageBytes: ByteArray?) : ChatItem
    data class ThoughtGroup(val durationSeconds: String, val skills: List<LoadedSkill>, val expanded: Boolean) : ChatItem
    data class AssistantText(val text: String) : ChatItem
    data class PromptEnhancement(val original: String, val enhanced: String) : ChatItem
    data class StatusBanner(val text: String) : ChatItem
    data class ResultImage(val bytes: ByteArray) : ChatItem
    data class Error(val message: String) : ChatItem
}

data class UiState(
    val items: List<ChatItem> = emptyList(),
    val busy: Boolean = false,
    val input: String = "",
    val attachedImage: ByteArray? = null,
    val attachedMime: String = "image/jpeg",
    val title: String = "Edit Image",
    val apiKeyConfigured: Boolean = false,
    val savedApiKey: String = "",
    val savedBaseUrl: String = "",
    val lastSaved: Boolean = false,
)

class EditImageViewModel(app: Application) : AndroidViewModel(app) {

    private val prefs = app.getSharedPreferences("agnes_settings", Application.MODE_PRIVATE)

    private var apiKey: String = prefs.getString("api_key", null) ?: BuildConfig.AGNES_API_KEY
    private var baseUrl: String =
        prefs.getString("base_url", null) ?: BuildConfig.AGNES_BASE_URL

    private var api = AgnesApi(apiKey, baseUrl)

    private val _state = MutableStateFlow(
        UiState(
            apiKeyConfigured = apiKey.isNotBlank(),
            savedApiKey = apiKey,
            savedBaseUrl = baseUrl,
        )
    )
    val state: StateFlow<UiState> = _state

    fun attachImage(uri: Uri) {
        val bytes = readBytes(getApplication(), uri)
        if (bytes.isNotEmpty()) {
            val mime = getApplication<Application>().contentResolver.getType(uri)
                ?: "image/jpeg"
            _state.update { it.copy(attachedImage = bytes, attachedMime = mime) }
        }
    }

    fun setInput(text: String) {
        _state.update { it.copy(input = text) }
    }

    fun clearAttachment() {
        _state.update { it.copy(attachedImage = null) }
    }

    fun reset() {
        _state.value = UiState(
            apiKeyConfigured = apiKey.isNotBlank(),
            savedApiKey = apiKey,
            savedBaseUrl = baseUrl,
        )
    }

    fun submit() {
        val s = _state.value
        val image = s.attachedImage ?: return
        val prompt = s.input.trim()
        if (prompt.isEmpty() || s.busy) return

        val base = s.items.size
        val items = s.items.toMutableList()
        val skillLoads = SKILLS.map { LoadedSkill(it.badge, it.content.length) }

        items += ChatItem.UserMessage(prompt, image)                 // base + 0
        items += ChatItem.ThoughtGroup("…", emptyList(), expanded = false)  // base + 1
        items += ChatItem.AssistantText("")                          // base + 2
        items += ChatItem.ThoughtGroup("…", skillLoads, expanded = true)    // base + 3
        items += ChatItem.PromptEnhancement(prompt, "")              // base + 4
        items += ChatItem.StatusBanner("Das dauert etwa 15–45 Sekunden, bitte habe einen Moment Geduld.") // base + 5

        _state.update {
            it.copy(
                items = items,
                busy = true,
                input = "",
                attachedImage = null,
                title = prompt,
                lastSaved = false,
            )
        }

        viewModelScope.launch {
            try {
                val imageDataUri = "data:${s.attachedMime};base64," +
                    Base64.encodeToString(image, Base64.NO_WRAP)

                val t0 = SystemClock.elapsedRealtime()
                val analysis = analyzeAndEnhance(api, imageDataUri, prompt)
                val analysisSeconds = (SystemClock.elapsedRealtime() - t0) / 1000.0

                _state.update { st ->
                    val newItems = st.items.toMutableList()
                    newItems[base + 1] = ChatItem.ThoughtGroup(
                        durationSeconds = String.format("%.2f", analysisSeconds),
                        skills = emptyList(),
                        expanded = false,
                    )
                    newItems[base + 2] = ChatItem.AssistantText(analysis.replyDe)
                    newItems[base + 4] = ChatItem.PromptEnhancement(prompt, analysis.editPrompt)
                    st.copy(items = newItems)
                }

                val t1 = SystemClock.elapsedRealtime()
                val dims = imageDimensions(image)
                val ratio = dims?.let { pickRatio(it.first, it.second) } ?: "3:4"
                val resultBytes = generateEdit(api, imageDataUri, analysis, ratio, "2K")
                val genSeconds = (SystemClock.elapsedRealtime() - t1) / 1000.0

                _state.update { st ->
                    val newItems = st.items.toMutableList()
                    newItems[base + 3] = ChatItem.ThoughtGroup(
                        durationSeconds = String.format("%.2f", genSeconds),
                        skills = skillLoads,
                        expanded = true,
                    )
                    newItems += ChatItem.AssistantText("Ich habe die gewünschte Änderung vorgenommen.")
                    newItems += ChatItem.ResultImage(resultBytes)
                    st.copy(items = newItems, busy = false)
                }
            } catch (e: Exception) {
                _state.update { st ->
                    val newItems = st.items.toMutableList()
                    newItems += ChatItem.Error(e.message ?: "Unbekannter Fehler")
                    st.copy(items = newItems, busy = false)
                }
            }
        }
    }

    fun updateSettings(apiKey: String, baseUrl: String) {
        val newKey = apiKey.trim()
        val newBase = baseUrl.trim().ifBlank { BuildConfig.AGNES_BASE_URL }
        this.apiKey = newKey
        this.baseUrl = newBase
        prefs.edit()
            .putString("api_key", newKey)
            .putString("base_url", newBase)
            .apply()
        api = AgnesApi(newKey, newBase)
        _state.update {
            it.copy(
                apiKeyConfigured = newKey.isNotBlank(),
                savedApiKey = newKey,
                savedBaseUrl = newBase,
            )
        }
    }

    fun markSaved() {
        _state.update { it.copy(lastSaved = true) }
    }
}

File: Windows/AgnesEditImage/ViewModels/EditImageViewModel.cs
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Input;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using AgnesEditImage.Models;
using AgnesEditImage.Services;

namespace AgnesEditImage.ViewModels;

public partial class EditImageViewModel : ObservableObject
{
    private readonly SettingsService _settingsService;
    private AgnesApi _api;

    [ObservableProperty]
    private ObservableCollection<ChatItem> _items = new();

    [ObservableProperty]
    private bool _busy;

    [ObservableProperty]
    private string _input = "";

    [ObservableProperty]
    private List<Attachment> _attachments = new();

    [ObservableProperty]
    private string _mode = "edit";

    [ObservableProperty]
    private string _title = "Edit Image";

    [ObservableProperty]
    private bool _apiKeyConfigured;

    [ObservableProperty]
    private string _savedApiKey = "";

    [ObservableProperty]
    private string _savedBaseUrl = "";

    [ObservableProperty]
    private bool _lastSaved;

    public EditImageViewModel()
    {
        _settingsService = new SettingsService();
        var settings = _settingsService.Load();
        _api = new AgnesApi(settings.ApiKey, settings.BaseUrl);
        _savedApiKey = settings.ApiKey;
        _savedBaseUrl = settings.BaseUrl;
        _apiKeyConfigured = !string.IsNullOrWhiteSpace(settings.ApiKey);
    }

    [RelayCommand]
    private void AddImages()
    {
        var dialog = new Microsoft.Win32.OpenFileDialog
        {
            Multiselect = true,
            Filter = "Bilder|*.jpg;*.jpeg;*.png;*.webp"
        };

        if (dialog.ShowDialog() != true) return;

        foreach (var file in dialog.FileNames.Take(10))
        {
            try
            {
                var bytes = File.ReadAllBytes(file);
                if (bytes.Length == 0) continue;
                var mime = Path.GetExtension(file).ToLowerInvariant() switch
                {
                    ".jpg" or ".jpeg" => "image/jpeg",
                    ".png" => "image/png",
                    ".webp" => "image/webp",
                    _ => "image/jpeg"
                };
                Attachments.Add(new LocalAttachment(bytes, mime));
            }
            catch
            {
                // ignore unreadable files
            }
        }
    }

    [RelayCommand]
    private void AddImageUrl()
    {
        var dialog = new UrlInputWindow();
        if (dialog.ShowDialog() != true) return;

        var url = dialog.EnteredUrl?.Trim();
        if (string.IsNullOrWhiteSpace(url)) return;

        if (Uri.TryCreate(url, UriKind.Absolute, out var uri) &&
            (uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps))
        {
            Attachments.Add(new RemoteAttachment(url));
        }
    }

    [RelayCommand]
    private void RemoveAttachment(int index)
    {
        if (index < 0 || index >= Attachments.Count) return;
        Attachments.RemoveAt(index);
    }

    [RelayCommand]
    private void UseAsInput(byte[] bytes)
    {
        if (bytes == null || bytes.Length == 0) return;
        Attachments = new List<Attachment> { new LocalAttachment(bytes, "image/png") };
    }

    [RelayCommand]
    private void SetMode(string mode)
    {
        Mode = mode;
    }

    [RelayCommand]
    private void Reset()
    {
        var settings = _settingsService.Load();
        Items = new ObservableCollection<ChatItem>();
        Busy = false;
        Input = "";
        Attachments = new List<Attachment>();
        Mode = "edit";
        Title = "Edit Image";
        ApiKeyConfigured = !string.IsNullOrWhiteSpace(settings.ApiKey);
        SavedApiKey = settings.ApiKey;
        SavedBaseUrl = settings.BaseUrl;
        LastSaved = false;
    }

    public void UpdateSettings(string apiKey, string baseUrl)
    {
        var newKey = apiKey?.Trim() ?? "";
        var newBase = baseUrl?.Trim() ?? "";
        if (string.IsNullOrWhiteSpace(newBase))
        {
            newBase = "https://apihub.agnes-ai.com/v1";
        }

        _settingsService.Save(newKey, newBase);
        _api = new AgnesApi(newKey, newBase);
        SavedApiKey = newKey;
        SavedBaseUrl = newBase;
        ApiKeyConfigured = !string.IsNullOrWhiteSpace(newKey);
        LastSaved = true;
    }

    [RelayCommand]
    private async Task Submit()
    {
        if (Attachments.Count == 0) return;
        var prompt = Input.Trim();
        if (prompt.Length == 0 || Busy) return;

        var images = Attachments.ToList();
        var skillLoads = Skills.All.Select(s => new LoadedSkill(s.Badge, s.Content.Length)).ToList();

        Items.Clear();
        Items.Add(new UserMessage(prompt, images));
        Items.Add(new ThoughtGroup("…", Skills: new List<LoadedSkill>(), Expanded: false));
        Items.Add(new ThoughtGroup("…", skillLoads, Expanded: true));
        Items.Add(new AssistantText(""));
        Items.Add(new PromptEnhancement(prompt, ""));
        Items.Add(new StatusBanner("Das dauert etwa 15–45 Sekunden, bitte habe einen Moment Geduld.", Active: true));

        Busy = true;
        Input = "";
        Attachments = new List<Attachment>();
        Title = prompt;

        try
        {
            var imageDataUris = images.Select(att =>
            {
                if (att is LocalAttachment local)
                {
                    var mime = string.IsNullOrWhiteSpace(local.Mime) ? "image/jpeg" : local.Mime;
                    return $"data:{mime};base64,{Convert.ToBase64String(local.Bytes)}";
                }
                return ((RemoteAttachment)att).Url;
            }).ToList();

            var t0 = Environment.TickCount;
            var analysis = await EditImagePipeline.AnalyzeAndEnhanceAsync(_api, imageDataUris, prompt);
            var analysisSeconds = (Environment.TickCount - t0) / 1000.0;

            Items[1] = new ThoughtGroup($"{analysisSeconds:F2}", Skills: new List<LoadedSkill>(), Expanded: false);
            Items[3] = new AssistantText(analysis.ReplyDe);
            Items[4] = new PromptEnhancement(prompt, analysis.EditPrompt);

            var t1 = Environment.TickCount;
            var firstLocal = images.OfType<LocalAttachment>().FirstOrDefault()?.Bytes;
            var dims = firstLocal != null ? ImageProcessor.ImageDimensions(firstLocal) : null;
            var ratio = dims.HasValue ? EditImagePipeline.PickRatio(dims.Value.Width, dims.Value.Height) : "3:4";
            var resultBytes = await EditImagePipeline.GenerateEditAsync(_api, imageDataUris, analysis, ratio, "2K", Mode);
            var genSeconds = (Environment.TickCount - t1) / 1000.0;

            Items[2] = new ThoughtGroup($"{genSeconds:F2}", skillLoads, Expanded: true);
            Items[5] = new StatusBanner("Bearbeitung abgeschlossen.", Active: false);
            Items.Add(new AssistantText("Ich habe die gewünschte Änderung vorgenommen."));
            Items.Add(new ResultImage(resultBytes));
        }
        catch (Exception ex)
        {
            if (Items.Count > 5)
            {
                Items[5] = new StatusBanner("Bearbeitung fehlgeschlagen.", Active: false);
            }
            Items.Add(new ErrorItem(ex.Message ?? "Unbekannter Fehler"));
        }
        finally
        {
            Busy = false;
        }
    }

    [RelayCommand]
    private void Download(byte[] bytes)
    {
        if (bytes == null || bytes.Length == 0) return;

        var path = ImageProcessor.SaveToGallery(bytes);
        if (path != null)
        {
            MessageBox.Show($"Bild gespeichert unter:\n{path}", "Gespeichert", MessageBoxButton.OK, MessageBoxImage.Information);
        }
        else
        {
            MessageBox.Show("Speichern fehlgeschlagen.", "Fehler", MessageBoxButton.OK, MessageBoxImage.Error);
        }
    }

    [RelayCommand]
    private void OpenFullscreen(byte[] bytes)
    {
        if (bytes == null || bytes.Length == 0) return;

        var bitmap = ImageProcessor.DecodeBitmap(bytes);
        if (bitmap == null) return;

        var window = new FullscreenWindow(bitmap);
        window.ShowDialog();
    }
}
